vivarium/simulator/states.py

```diff
@@ -0,0 +1,316 @@
+
+    def __getattr__(self, name):
+        def wrapper(e_type):
+            value = getattr(self.nve_state, name)
+            if isinstance(value, rigid_body.RigidBody):
+                return rigid_body.RigidBody(center=value.center[self.e_cond(e_type)],
+                                            orientation=value.orientation[self.e_cond(e_type)])
+            else:
+                return value[self.e_cond(e_type)]
+        return wrapper
+
+
+# Helper function to transform a color string into rgb with matplotlib colors
+def _string_to_rgb(color_str):
+    return jnp.array(list(mcolors.to_rgb(color_str)))
+
+
+def init_simulator_state(
+        box_size: float = 100.,
+        n_agents: int = 10,
+        n_objects: int = 2,
+        num_steps_lax: int = 4,
+        dt: float = 0.1,
+        freq: float = 40.,
+        neighbor_radius: float = 100.,
+        to_jit: bool = True,
+        use_fori_loop: bool = False
+        ) -> SimulatorState:
+    """
+    Initialize simulator state with given parameters
+    """
+    return SimulatorState(
+        idx=jnp.array([0]),
+        box_size=jnp.array([box_size]),                       
+        n_agents=jnp.array([n_agents]),
+        n_objects=jnp.array([n_objects]),
+        num_steps_lax=jnp.array([num_steps_lax], dtype=int),
+        dt=jnp.array([dt], dtype=float),
+        freq=jnp.array([freq], dtype=float),
+        neighbor_radius=jnp.array([neighbor_radius], dtype=float),
+        # Use 1*bool to transform True to 1 and False to 0
+        to_jit= jnp.array([1*to_jit]),
+        use_fori_loop=jnp.array([1*use_fori_loop]))
+
+
+def _init_positions(key_pos, positions, n_elements, box_size, n_dims=2):
+    assert (len(positions) == n_elements if positions else True)
+    # If positions are passed, transform them in jax array
+    if positions:
+        positions = jnp.array(positions)
+    # Else initialize random positions
+    else:
+        positions = random.uniform(key_pos, (n_elements, n_dims)) * box_size
+    return positions
+
+def _init_existing(n_existing, n_elements):
+    if n_existing:
+        assert n_existing <= n_elements
+        existing_arr = jnp.ones((n_existing))
+        non_existing_arr = jnp.zeros((n_elements - n_existing))
+        exists_array = jnp.concatenate((existing_arr, non_existing_arr))
+    else:
+        exists_array = jnp.ones((n_elements))
+    return exists_array
+
+# TODO : Should also add union float, list[float] for friction, diameter ..
+# Also think it would be easier to only handle the nve state in the physics engine. (if it doesn't make the  simulation run slower)
+# Here it makes a really long function, and it isn't very modular 
+# Plus we store different attributes of entities (e.g Agents positions and sensors) in different dataclasses
+def init_nve_state(
+        simulator_state: SimulatorState,
+        diameter: float = 5.,
+        friction: float = 0.1,
+        mass_center: float = 1.,
+        mass_orientation: float = 0.125,
```

Comment:
We should be able to pass a full array as well here, e.g. if we want to set different diameters for different entities. A solution could be to check the type of this argument : if it is a scalar, all entities will have the same value, if it is an array, we set them as such. Will require to explain it in a docstring ;)